{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc7d086d-a6d5-4337-b2de-1332fcbe9e97",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Modeling the dynamics of EMT reveals genes associated with pan-cancer intermediate states and plasticity\n",
    "\n",
    "The bioRxiv preprint is available [here](https://www.biorxiv.org/content/10.1101/2024.10.03.616309v1). \n",
    "\n",
    "\n",
    "### Overview \n",
    "\n",
    "This repository contains code used to analyse the epithelial-to-mesenchymal transition (EMT) using single-cell RNA sequencing (scRNA-seq) data. The code is written in Python (for single-cell processing pipeline), Julia (for modeling and Bayesian parameter inference), and R (for cell scoring). Code is formatted in Jupyter Lab / Jupyter notebook (.ipynb) and R markdown (.Rmd) filetypes.\n",
    "\n",
    "\n",
    "## Data\n",
    "\n",
    "Raw data used for the analysis is from public repositories:\n",
    "* Cook and Vanderhyden, 2020 (GSE147405)\n",
    "* Pastushenko et al., 2018 (GEO accession GSE110357)\n",
    "* van Dijk et al., 2018 (GSE114397)\n",
    "* Panchy et al., 2022 (GSE213753)\n",
    "(add full citations later)\n",
    "\n",
    "\n",
    "## Project contents\n",
    "\n",
    "1 - scanpy\n",
    "\n",
    "scRNAseq processing pipeline, primarily based on Scanpy code.\n",
    "\n",
    "Each notebook is designed to run a single sample from the titular dataset. The sample can be changed at the beginning and will import sample-specific parameters from all_run_settings.xlsx and functions from universal_dataset_functions.ipynb \\& the titular functions.ipynb.\n",
    "\n",
    "universal_dataset_functions.ipynb contains functions for all samples.\n",
    "All \\*\\_functions.ipynb files contain overriding functions that are specific to each dataset.\n",
    "\n",
    "The code for each sample can be run with:\n",
    "* 3 - Pastushenko_singleRun.ipynb\n",
    "* 5 - vanDijk_singleRun.ipynb\n",
    "* 7 - Cook_singleRun.ipynb\n",
    "* 9 - Panchy.ipynb\n",
    "with no further adjustments to other files, as all other files are imported (with exception for directory references).\n",
    "\n",
    "\n",
    "Directory contains:\n",
    "\n",
    "1 - universal_dataset_functions.ipynb\n",
    "2 - Pastushenko_functions.ipynb\n",
    "3 - Pastushenko_singleRun.ipynb\n",
    "4 - vanDijk_functions.ipynb\n",
    "5 - vanDijk_singleRun.ipynb\n",
    "6 - Cook_functions.ipynb\n",
    "7 - Cook_singleRun.ipynb\n",
    "8 - Cook_singleRun_removedRuns.ipynb\n",
    "9 - Panchy.ipynb\n",
    "9.5 - Panchy_getKRT15highcells.ipynb\n",
    "9.5 - Panchy_KRT15highcells.csv\n",
    "all_run_settings.xlsx\n",
    "\n",
    "\n",
    "2 - ucell code\n",
    "\n",
    "EMTscore - removed samples.Rmd\n",
    "EMTscore - samples.Rmd\n",
    "\n",
    "3 - DE genes\n",
    "\n",
    "Epithelial - log2FC analysis.ipynb\n",
    "Intermediate - log2FC analysis.ipynb\n",
    "Mesenchymal - log2FC analysis.ipynb\n",
    "nCells directory\n",
    "\n",
    "4 - rnavelocity code\n",
    "\n",
    "\\_withoutContaminantCellLine_OVCA420-EGF.csv\n",
    "\\_withoutContaminantCellLine_OVCA420-TGFB1.csv\n",
    "\\_withoutContaminantCellLine_OVCA420-TNF.csv\n",
    "1 - Cook_singleRun_filterContaminantCells_OVCA420.ipynb\n",
    "2 - Cook_singleRun_realigned.ipynb\n",
    "Cook-realigned_run_settings.xlsx\n",
    "\n",
    "5 - ode model code\n",
    "\n",
    "Bayesian parameter fitting, one I state.ipynb\n",
    "Bayesian parameter fitting, two I state.ipynb\n",
    "graphing fitted parameters, one I state.ipynb\n",
    "graphing fitted parameters, two I state.ipynb\n",
    "\n",
    "6 - ODE genes\n",
    "\n",
    "Epithelial - kparam correlations.ipynb\n",
    "Intermediate - kparam correlations.ipynb\n",
    "Mesenchymal - kparam correlations.ipynb\n",
    "nCells directory\n",
    "\n",
    "marker gene lists\n",
    "\n",
    "Cell Cycle Markers - G1,S Genes.csv\n",
    "Cell Cycle Markers - G2,M Genes.csv\n",
    "\n",
    "Markers - EMP Cook 2022.csv\n",
    "Markers - MSigDB.csv\n",
    "Markers - PanglaoDB, Epithelial.csv\n",
    "\n",
    "\n",
    "## Package requirements\n",
    "\n",
    "Conda environments are available in the directory environment files.\n",
    "\n",
    "\n",
    "\n",
    "### Package requirements \n",
    " - [DifferentialEquations.jl](https://diffeq.sciml.ai/stable/)\n",
    " - [Turing.jl](https://turing.ml/stable/)\n",
    " - MCMCChains.jl\n",
    " - StochasticDelayDiffEq.jl\n",
    " - DiffEqSensitivity.jl\n",
    " - DiffEqCallbacks.jl\n",
    " - DecisionTree.jl\n",
    " - ScikitLearn.CrossValidation\n",
    " - ModelingToolkit.jl\n",
    " - StatsPlots.jl\n",
    " - Distributions.jl\n",
    " - Statistics.jl\n",
    " - Catalyst.jl\n",
    " - ParameterizedFunctions.jl\n",
    " - DiffeqJump.jl\n",
    " - Plots.jl; pyplot()\n",
    " - DataFrames.jl\n",
    " - DelimitedFiles.jl\n",
    " - CSV.jl\n",
    " - JLD.jl\n",
    "\n",
    "### Project contents\n",
    " - `README.md` : this file with information about the repository and [paper](https://doi.org/10.1101/2022.06.15.496246)\n",
    " - `Modeling_MDSCs.ipynb` :  Jupyter notebook containing code blocks for all simulations and figures in the paper. Code blocks within the notebook are intended to be run independently. \n",
    " - `Modeling_MDSCs_julia_v1.8.ipynb` :  Jupyter notebook (updated for Julia 1.8) containing code blocks for all simulations and figures in the paper. Code blocks within the notebook are intended to be run independently. \n",
    " - `tumor_data.xlsx` : data used in the analysis, see Spigel, D. R. et al. FIR: Efficacy, Safety, and Biomarker Analysis of a Phase II Open-Label Study of Atezolizumab in PD-L1–Selected Patients With NSCLC. Journal of Thoracic Oncology 13, 1733–1742 (2018). URL [https://www.jto.org/article/S1556-0864(18)30603-8/fulltext](https://www.jto.org/article/S1556-0864(18)30603-8/fulltext) and Laleh, N. G. et al. Classical mathematical models for prediction of response to chemotherapy and immunotherapy. PLOS Computational Biology 18, e1009822 (2022). URL [https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009822](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009822).\n",
    " - `gillespie.csv` : Gillespie simulation results (see Figure S5)\n",
    "\n",
    "### Acknowledgments\n",
    "We would like to thank E.J. Fertig for valuable discussions and guidance, and the Tumor Microenvironment Program at the USC-Norris Comprehensive Cancer Center for their support. We would like to thank all members of the Roussos Torres and MacLean labs for valuable input and discussions. Figures 1A, 5A, & 6A were created with [BioRender](https://biorender.com/).\n",
    "\n",
    "### Contributors\n",
    "<a href=\"https://github.com/maclean-lab/ModelingMDSCs/graphs/contributors\">\n",
    "  <img src=\"https://contributors-img.web.app/image?repo=maclean-lab/ModelingMDSCs\" />\n",
    "</a>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8ec504c-ea87-4d92-8385-3b89ac3769d1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
